### PR TITLE
real async _readTexturePixels

### PR DESCRIPTION
**Pull request by RolandCsibrei**
https://forum.babylonjs.com/t/why-async-readtexturepixels-uses-synchronous-version-of-readtexturepixels/51290

### CODE REVIEW

**Comment by bjsplat** [comment]
Please make sure to label your PR with "bug", "new feature" or "breaking change" label(s).
To prevent this PR from going to the changelog marked it with the "skip changelog" label.

**Comment by bjsplat** [comment]
Snapshot stored with reference name:
refs/pull/15312/merge

Test environment: 
https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/15312/merge/index.html

To test a playground add it to the URL, for example:

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/15312/merge/index.html#WGZLGJ#4600

Links to test babylon tools with this snapshot:

https://playground.babylonjs.com/?snapshot=refs/pull/15312/merge
https://sandbox.babylonjs.com/?snapshot=refs/pull/15312/merge
https://gui.babylonjs.com/?snapshot=refs/pull/15312/merge
https://nme.babylonjs.com/?snapshot=refs/pull/15312/merge

To test the snapshot in the playground with a playground ID add it after the snapshot query string:

https://playground.babylonjs.com/?snapshot=refs/pull/15312/merge#BCU1XR#0

**Comment by bjsplat** [comment]
WebGL2 visualization test reporter:

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/15312/merge/testResults/webgl2playwright/index.html

**Comment by bjsplat** [comment]
Visualization tests for WebGPU (Experimental)

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/15312/merge/testResults/webgpuplaywright/index.html

**Comment by RolandCsibrei** [comment]
https://forum.babylonjs.com/t/why-async-readtexturepixels-uses-synchronous-version-of-readtexturepixels/51290

File: packages/dev/core/src/Engines/engine.ts [hunk 1]
```diff
@@ -990,17 +990,17 @@ export class Engine extends ThinEngine {
...
-    private _clientWaitAsync(sync: WebGLSync, flags = 0, intervalms = 10): Promise<void> {
+    private _clientWaitAsync(sync: WebGLSync, flags = 0, checkForSyncInterval = 10): Promise<void> {
         const gl = <WebGL2RenderingContext>(this._gl as any);
         return new Promise((resolve, reject) => {
             const check = () => {
-                const res = gl.clientWaitSync(sync, flags, 0);
+                const res = gl.clientWaitSync(sync, flags, 0); // timeout = 0, checks SYNC state without blocking
                 if (res == gl.WAIT_FAILED) {
-                    reject();
+                    reject("WebGL sync object failed");
```
Comment (Taras55) [inline packages/dev/core/src/Engines/engine.ts:999]: reject(new Error("WebGL sync object failed"));

Comment (RolandCsibrei) [reply packages/dev/core/src/Engines/engine.ts]: Not really needed but let's have the stacktrace logged as well. 

Changed.

File: packages/dev/core/src/Engines/Extensions/engine.readTexture.ts [hunk 2]
```diff
@@ -2,8 +2,12 @@ import { ThinEngine } from "../../Engines/thinEngine";
 import type { InternalTexture } from "../../Materials/Textures/internalTexture";
 import type { Nullable } from "../../types";
 import { Constants } from "../constants";
+import type { Engine } from "../engine";
 
 declare module "../../Engines/abstractEngine" {
+    /**
```
Comment (sebavan) [inline packages/dev/core/src/Engines/Extensions/engine.readTexture.ts:8]: this should be removed

File: packages/dev/core/src/Engines/Extensions/engine.readTexture.ts [hunk 3]
```diff
@@ -167,16 +170,35 @@ ThinEngine.prototype._readTexturePixelsSync = function (
     }
 
     if (flushRenderer) {
-        this.flushFramebuffer();
+        engine.flushFramebuffer();
     }
 
-    gl.readPixels(x, y, width, height, gl.RGBA, readType, <DataView>buffer);
+    return { outputBuffer: buffer, readType };
```
Comment (sebavan) [inline packages/dev/core/src/Engines/Extensions/engine.readTexture.ts:176]: can we avoid creating an object here for better GC

**Review by sebavan — CHANGES_REQUESTED** [review]
Would be great to validate how faster it is in the scenario we use and read textures

File: packages/dev/core/src/Engines/Extensions/engine.readTexture.ts (hunk 3 above)
Comment (RolandCsibrei) [reply packages/dev/core/src/Engines/Extensions/engine.readTexture.ts:176]: Sure.

**Comment by RolandCsibrei** [comment]
I will reopen a new one.

**Comment by bjsplat** [comment]
Visualization tests for WebGL 1 have failed. If some tests failed because the snapshots do not match, the report can be found at

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/15312/merge/testResults/webgl1/index.html

If tests were successful afterwards, this report might not be available anymore.

**Pull request closed without merging**